File: webapp/lib/volc-signer.ts
/**
 * 火山引擎 Volcengine Visual API V4 HMAC-SHA256 签名 (TypeScript port).
 *
 * 端口自 src/common/volc_signer.py (Python production-verified).
 *
 * 锁定:
 *   - 算法: HMAC-SHA256
 *   - 区域: cn-north-1
 *   - 服务: cv
 *   - Host: visual.volcengineapi.com
 *   - 签名头: content-type;host;x-content-sha256;x-date
 *   - Credential scope: <yyyymmdd>/cn-north-1/cv/request
 *
 * 使用 Node.js 18+ 原生 crypto，零依赖。
 */
import crypto from 'node:crypto';

const REGION = 'cn-north-1';
const SERVICE = 'cv';
const HOST = 'visual.volcengineapi.com';
const ALGORITHM = 'HMAC-SHA256';

export interface SignedRequest {
  method: 'POST';
  url: string;
  headers: Record<string, string>;
  body: Buffer;
}

function sha256Hex(data: Buffer | string): string {
  return crypto.createHash('sha256').update(data).digest('hex');
}

function hmacSha256(key: Buffer | string, msg: string): Buffer {
  return crypto.createHmac('sha256', key).update(msg, 'utf8').digest();
}

function deriveSigningKey(
  secretKey: string,
  datestamp: string,
  region: string,
  service: string,
): Buffer {
  const kDate = hmacSha256(secretKey, datestamp);
  const kRegion = hmacSha256(kDate, region);
  const kService = hmacSha256(kRegion, service);
  return hmacSha256(kService, 'request');
}

function rfc3986EncodeURIComponent(str: string): string {
  // V4 signing requires RFC 3986 percent-encoding (`!*'()` must be encoded;
  // `-_.~` are safe). JS encodeURIComponent leaves `!*'()` unencoded, so patch.
  return encodeURIComponent(str).replace(
    /[!*'()]/g,
    (c) => '%' + c.charCodeAt(0).toString(16).toUpperCase(),
  );
}

function canonicalQueryString(params: Record<string, string>): string {
  const keys = Object.keys(params).sort();
  if (keys.length === 0) return '';
  return keys
    .map((k) => `${rfc3986EncodeURIComponent(k)}=${rfc3986EncodeURIComponent(params[k])}`)
    .join('&');
}

export interface SignRequestOpts {
  accessKey: string;
  secretKey: string;
  action: string;
  version: string;
  body: Buffer | string;
  extraQuery?: Record<string, string>;
}

export function signRequest(opts: SignRequestOpts): SignedRequest {
  const bodyBuf = Buffer.isBuffer(opts.body) ? opts.body : Buffer.from(opts.body, 'utf8');

  const now = new Date();
  const yyyy = now.getUTCFullYear().toString().padStart(4, '0');
  const mm = (now.getUTCMonth() + 1).toString().padStart(2, '0');
  const dd = now.getUTCDate().toString().padStart(2, '0');
  const HH = now.getUTCHours().toString().padStart(2, '0');
  const MM = now.getUTCMinutes().toString().padStart(2, '0');
  const SS = now.getUTCSeconds().toString().padStart(2, '0');
  const amzdate = `${yyyy}${mm}${dd}T${HH}${MM}${SS}Z`;
  const datestamp = `${yyyy}${mm}${dd}`;

  const query: Record<string, string> = { Action: opts.action, Version: opts.version };
  if (opts.extraQuery) {
    for (const [k, v] of Object.entries(opts.extraQuery)) {
      if (v !== undefined && v !== null) query[k] = String(v);
    }
  }
  const canonicalQS = canonicalQueryString(query);

  const payloadHash = sha256Hex(bodyBuf);

  const canonicalHeaders =
    'content-type:application/json\n' +
    `host:${HOST}\n` +
    `x-content-sha256:${payloadHash}\n` +
    `x-date:${amzdate}\n`;
  const signedHeaders = 'content-type;host;x-content-sha256;x-date';

  const canonicalRequest = [
    'POST',
    '/',
    canonicalQS,
    canonicalHeaders,
    signedHeaders,
    payloadHash,
  ].join('\n');

  const credentialScope = `${datestamp}/${REGION}/${SERVICE}/request`;
  const stringToSign = [
    ALGORITHM,
    amzdate,
    credentialScope,
    sha256Hex(canonicalRequest),
  ].join('\n');

  const signingKey = deriveSigningKey(opts.secretKey, datestamp, REGION, SERVICE);
  const signature = crypto.createHmac('sha256', signingKey).update(stringToSign, 'utf8').digest('hex');

  const authorization =
    `${ALGORITHM} ` +
    `Credential=${opts.accessKey}/${credentialScope}, ` +
    `SignedHeaders=${signedHeaders}, ` +
    `Signature=${signature}`;

  return {
    method: 'POST',
    url: `https://${HOST}/?${canonicalQS}`,
    headers: {
      'Content-Type': 'application/json',
      Host: HOST,
      'X-Date': amzdate,
      'X-Content-Sha256': payloadHash,
      Authorization: authorization,
    },
    body: bodyBuf,
  };
}

File: src/common/volc_signer.py
"""火山引擎 Volcengine Visual API V4 HMAC-SHA256 签名实现.

锁定证据：基于 GitLab self-media-james/即梦.md (2026-03-12 更新) 的实测可跑通版本，
所有签名字段、规范化请求、credential_scope 公式均已实测验证。
"""
from __future__ import annotations

import datetime as _dt
import hashlib
import hmac
from dataclasses import dataclass
from typing import Mapping
from urllib.parse import quote


_REGION = "cn-north-1"
_SERVICE = "cv"
_HOST = "visual.volcengineapi.com"
_ALGORITHM = "HMAC-SHA256"


def _sign(key: bytes, msg: str) -> bytes:
    return hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()


def _derive_signing_key(secret_key: str, datestamp: str, region: str, service: str) -> bytes:
    k_date = _sign(secret_key.encode("utf-8"), datestamp)
    k_region = _sign(k_date, region)
    k_service = _sign(k_region, service)
    return _sign(k_service, "request")


def _canonical_query_string(params: Mapping[str, str]) -> str:
    if not params:
        return ""
    pairs = sorted(params.items())
    return "&".join(
        f"{quote(str(k), safe='-_.~')}={quote(str(v), safe='-_.~')}" for k, v in pairs
    )


@dataclass(frozen=True)
class SignedRequest:
    method: str
    url: str
    headers: dict
    body: bytes


def sign_request(
    *,
    access_key: str,
    secret_key: str,
    action: str,
    version: str,
    body: bytes,
    method: str = "POST",
    region: str = _REGION,
    service: str = _SERVICE,
    host: str = _HOST,
    extra_query: Mapping[str, str] | None = None,
) -> SignedRequest:
    """Sign a Volcengine OpenAPI request with V4 HMAC-SHA256."""

    now = _dt.datetime.now(_dt.timezone.utc)
    amzdate = now.strftime("%Y%m%dT%H%M%SZ")
    datestamp = now.strftime("%Y%m%d")

    query = {"Action": action, "Version": version}
    if extra_query:
        query.update({k: v for k, v in extra_query.items() if v is not None})
    canonical_qs = _canonical_query_string(query)

    payload_hash = hashlib.sha256(body or b"").hexdigest()

    # NOTE: Volc vefaas / apig OpenAPI reject signatures whose canonical
    # content-type omits the charset segment (returns
    # InvalidAuthorization 100024). Keep `application/json; charset=utf-8`
    # to stay compatible with both visual.volcengineapi.com and
    # open.volcengineapi.com.
    canonical_headers = (
        "content-type:application/json; charset=utf-8\n"
        f"host:{host}\n"
        f"x-content-sha256:{payload_hash}\n"
        f"x-date:{amzdate}\n"
    )
    signed_headers = "content-type;host;x-content-sha256;x-date"

    canonical_request = "\n".join([
        method.upper(),
        "/",
        canonical_qs,
        canonical_headers,
        signed_headers,
        payload_hash,
    ])

    credential_scope = f"{datestamp}/{region}/{service}/request"
    string_to_sign = "\n".join([
        _ALGORITHM,
        amzdate,
        credential_scope,
        hashlib.sha256(canonical_request.encode("utf-8")).hexdigest(),
    ])

    signing_key = _derive_signing_key(secret_key, datestamp, region, service)
    signature = hmac.new(signing_key, string_to_sign.encode("utf-8"), hashlib.sha256).hexdigest()

    authorization = (
        f"{_ALGORITHM} "
        f"Credential={access_key}/{credential_scope}, "
        f"SignedHeaders={signed_headers}, "
        f"Signature={signature}"
    )

    return SignedRequest(
        method=method.upper(),
        url=f"https://{host}/?{canonical_qs}",
        headers={
            # MUST match canonical_headers exactly (incl. ; charset=utf-8)
            "Content-Type": "application/json; charset=utf-8",
            "Host": host,
            "X-Date": amzdate,
            "X-Content-Sha256": payload_hash,
            "Authorization": authorization,
        },
        body=body,
    )
